### PR TITLE
Check if device revoked or time expired after couple of failed attempts

### PR DESCRIPTION
**Pull request by sajacl**
Counting number of failed attempts, and after an interval, trying to diagnostic the error:

1. Check if device is revoked,
2. Check if account ran out of time
3. Report error

This change is [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4113)

### CODE REVIEW

File: ios/MullvadREST/RESTNetworkOperation.swift [hunk 1]
```diff
@@ -281,8 +281,23 @@ extension REST {
             // Increment retry count.
             retryCount += 1
 
-            // Retry immediatly if retry delay is set to never.
-            guard retryStrategy.retryDelay != .never else {
+            switch retryStrategy.retryDelay {
```
Comment (pronebird) [inline ios/MullvadREST/RESTNetworkOperation.swift:284]: `WaitStrategy` could produce iterator for each specific strategy. Then operation wouldn't have to account for each specific stragedy.

File: ios/PacketTunnel/PacketTunnelProvider.swift [hunk 2]
```diff
@@ -307,6 +322,144 @@ class PacketTunnelProvider: NEPacketTunnelProvider, TunnelMonitorDelegate {
...
+                    completion(.ranOutOfTime(newExpiry: accountData.expiry))
+                }
+            }
+
+            if let deviceCompletionError = deviceCompletion?.error {
+                completion(.fetchingDeviceDataFinishedWith(error: deviceCompletionError))
+            }
+        }
+
+        completionOperation.addDependencies([accountOperation, deviceOperation])
```
Comment (pronebird) [inline ios/PacketTunnel/PacketTunnelProvider.swift:404]: Forgot to add operations to `AsyncOperationQueue`?

File: ios/PacketTunnel/PacketTunnelProvider.swift [hunk 3]
```diff
@@ -307,6 +322,144 @@ class PacketTunnelProvider: NEPacketTunnelProvider, TunnelMonitorDelegate {
...
+        }
+
+        return operation
+    }
+
+    private func createGetDeviceDataOperation(accountNumber: String, identifier: String)
+        -> ResultOperation<REST.Device, REST.Error>
+    {
+        let dispatchQueue = DispatchQueue(label: "PacketTunnel.DeviceDataProxy", qos: .utility)
+        let operation = ResultBlockOperation<REST.Device, REST.Error>(dispatchQueue: dispatchQueue)
```
Comment (pronebird) [inline ios/PacketTunnel/PacketTunnelProvider.swift:359]: Same here, just pass `self.dispatchQueue`

File: ios/PacketTunnel/PacketTunnelProvider.swift [hunk 4]
```diff
@@ -307,6 +322,144 @@ class PacketTunnelProvider: NEPacketTunnelProvider, TunnelMonitorDelegate {
...
+        case fetchingAccountDataFinishedWith(error: Error)
+        case fetchingDeviceDataFinishedWith(error: REST.Error)
+        case ranOutOfTime(newExpiry: Date)
+    }
+
+    private func createGetAccountDataOperation(accountNumber: String)
+        -> ResultOperation<REST.AccountData, REST.Error>
+    {
+        let operation = ResultBlockOperation<REST.AccountData, REST.Error>(
+            dispatchQueue: DispatchQueue(label: "PacketTunnel.AccountDataProxy", qos: .utility)
```
Comment (pronebird) [inline ios/PacketTunnel/PacketTunnelProvider.swift:336]: Pass `self.dispatchQueue` queue.

Comment (sajacl) [reply ios/PacketTunnel/PacketTunnelProvider.swift]: Done.

File: ios/PacketTunnel/PacketTunnelProvider.swift (hunk 3 above)
Comment (sajacl) [reply ios/PacketTunnel/PacketTunnelProvider.swift:368]: Done.

File: ios/PacketTunnel/PacketTunnelProvider.swift [hunk 5]
```diff
@@ -307,6 +322,144 @@ class PacketTunnelProvider: NEPacketTunnelProvider, TunnelMonitorDelegate {
...
+        deviceIdentifier: String,
+        completion: @escaping (FailedToEstablishConnectionError?) -> Void
+    ) {
+        let accountOperation = createGetAccountDataOperation(accountNumber: accountNumber)
+        let deviceOperation = createGetDeviceDataOperation(
+            accountNumber: accountNumber,
+            identifier: deviceIdentifier
+        )
+
+        let completionOperation = AsyncBlockOperation(dispatchQueue: .global()) {
```
[the remark this answers is not in the document]
Comment (sajacl) [reply ios/PacketTunnel/PacketTunnelProvider.swift]: Done.

File: ios/PacketTunnel/PacketTunnelProvider.swift (hunk 2 above)
Comment (sajacl) [reply ios/PacketTunnel/PacketTunnelProvider.swift:413]: Done.

File: ios/MullvadREST/RESTNetworkOperation.swift (hunk 1 above)
Comment (sajacl) [reply ios/MullvadREST/RESTNetworkOperation.swift]: Done.

**Comment by pronebird** [comment]
Moved over to https://github.com/mullvad/mullvadvpn-app/pull/4124

**Pull request closed without merging**